L1Trigger/L1THGCal/plugins/backend/HGCalLayer1PhiOrderWrapper.cc

```diff
@@ -0,0 +1,126 @@
+
+  void configure(
+      const std::tuple<const HGCalTriggerGeometryBase* const, const unsigned&, const uint32_t&>& configuration) override;
+
+  void process(const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs,
+               l1t::HGCalClusterBxCollection& clusters) const override;
+
+private:
+  void convertCMSSWInputs(const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs,
+                          l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_SA) const;
+
+  void convertAlgorithmOutputs(const l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_out,
+                               const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs_original,
+                               l1t::HGCalClusterBxCollection& clusters) const;
+
+  void setGeometry(const HGCalTriggerGeometryBase* const geom) { triggerTools_.setGeometry(geom); }
+
+  double rotatedphi(double phi, unsigned sector) const;
+
+  HGCalTriggerTools triggerTools_;
+  l1thgcfirmware::HGCalLayer1PhiOrderFwImpl theAlgo_;
+  l1thgcfirmware::HGCalLayer1PhiOrderFwConfig theConfiguration_;
+
+  // Scale factor for quantities sent to emulator to keep floating point precision. Value is arbitrary and could be set to relevant value.
+  const unsigned int FWfactor_ = 10000;
+};
+
+HGCalLayer1PhiOrderWrapper::HGCalLayer1PhiOrderWrapper(const edm::ParameterSet& conf)
+    : HGCalLayer1PhiOrderWrapperBase(conf), theAlgo_(), theConfiguration_() {}
+
+void HGCalLayer1PhiOrderWrapper::convertCMSSWInputs(const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs,
+                                                    l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_SA) const {
+  fpga_tcs_SA.clear();
+  fpga_tcs_SA.reserve(fpga_tcs.size());
+  unsigned int itc = 0;
+  for (auto& tc : fpga_tcs) {
+    double phi = rotatedphi(tc->phi(), theConfiguration_.phiSector());
+    phi += (phi < 0) ? M_PI : 0;
+    unsigned int digi_phi = phi * FWfactor_;
+    unsigned int digi_energy = (tc->mipPt()) * FWfactor_;
+    fpga_tcs_SA.emplace_back(true,
+                             true,
+                             1,
+                             digi_phi,
+                             triggerTools_.layerWithOffset(tc->detId()),
+                             digi_energy);  //Assign dummy RoverZbin to not change the dataformat/creators.
+    fpga_tcs_SA.back().setModuleId(triggerTools_.getModuleId(tc->detId()));
+    fpga_tcs_SA.back().setCmsswIndex(std::make_pair(itc, 0));
+    ++itc;
+  }
+}
+
+void HGCalLayer1PhiOrderWrapper::convertAlgorithmOutputs(
+    const l1thgcfirmware::HGCalTriggerCellSACollection& fpga_tcs_out,
+    const std::vector<edm::Ptr<l1t::HGCalTriggerCell>>& fpga_tcs_original,
+    l1t::HGCalClusterBxCollection& clusters) const {
+  std::vector<l1t::HGCalCluster> clustersTmp;
+
+  for (std::vector<edm::Ptr<l1t::HGCalTriggerCell>>::const_iterator otc = fpga_tcs_original.begin();
+       otc != fpga_tcs_original.end();
+       ++otc) {
```

Comment:
It would be more compact to use a range-based for loop:
```suggestion
  for (const auto& otc : fpga_tcs_original) {
```